### PR TITLE
Return bit-rot verified data instead of re-reading from disk

### PR DESCRIPTION
**Pull request by donatello**
## Description

- Data from disk was being read after bitrot verification to return
      data for GetObject. Strictly speaking this does not guarantee bitrot
      protection, as disks may return bad data even temporarily.
    
- This fix reads data from disk, verifies data for bitrot and then
      returns data to the client directly.

## Motivation and Context

To return only bitrot verified data this PR reads full part-file shards into memory on the client, then reconstructs the data from erasure coding and returns data for GetObject.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#373cccd7-94ae-420e-ae20-8127261d058b) |
<a name="373cccd7-94ae-420e-ae20-8127261d058b"></a>
### 5568-4be6582/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30658
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp b3ecf80cf80a:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-3m1isu6gfc9j34",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 312,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:30658/aws-sdk-go-test-3m1isu6gfc9j34: dial tcp 192.168.122.45:30658: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#78a1829f-7740-4912-bb16-e7ae01069812) |
<a name="78a1829f-7740-4912-bb16-e7ae01069812"></a>
### 5568-475097c/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:31644
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp d79c78411b10:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-5a0kqndwa66rmt",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 256,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:31644/aws-sdk-go-test-5a0kqndwa66rmt: dial tcp 192.168.122.45:31644: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-fs.sh | :x: [more...](#760db24e-9bbd-4a51-948d-fcf56f5c4c1c) |
| mint-dist-xl.sh | :x: [more...](#ef9d88c2-86c4-4168-9a3e-9ba266487300) |
<a name="ef9d88c2-86c4-4168-9a3e-9ba266487300"></a>
### 5568-ef0bdfa/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:32205
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp ac88d6c1dacf:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-wpcuo64e2c39gx",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 306,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:32205/aws-sdk-go-test-wpcuo64e2c39gx: dial tcp 192.168.122.45:32205: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```
<a name="760db24e-9bbd-4a51-948d-fcf56f5c4c1c"></a>
### 5568-ef0bdfa/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:31945
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp dcb7e0848af4:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 0 seconds
(2/12) Running aws-sdk-php tests ... done in 48 seconds
(3/12) Running aws-sdk-ruby tests ... done in 11 seconds
(4/12) Running awscli tests ... done in 4 minutes and 51 seconds
(5/12) Running mc tests ... done in 2 minutes and 3 seconds
(6/12) Running minio-dotnet tests ... done in 49 seconds
(7/12) Running minio-go tests ... done in 2 seconds
(8/12) Running minio-java tests ... done in 55 seconds
(9/12) Running minio-js tests ... FAILED in 30 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook",
  "duration": 931,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:56:11) at node_modules/minio/dist/main/transformers.js:117:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:48:26) at DestroyableTransform.<anonymous> (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:135:12) at finishMaybe (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:371:12) at endWritable (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:378:3) at DestroyableTransform.Writable.end (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:356:5) at IncomingMessage.onend (_stream_readable.js:511:10) at endReadableNT (_stream_readable.js:974:12) at _combinedTickCallback (internal/process/next_tick.js:80:11) at process._tickCallback (internal/process/next_tick.js:104:9)"
}

Executed 8 out of 12 tests successfully.
```

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#abeef753-2a37-4fd7-b139-88e74b5e6472) |
<a name="abeef753-2a37-4fd7-b139-88e74b5e6472"></a>
### 5568-e7eedaf/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30522
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 456d60f42100:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-m3t66mrhlzlb16",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 399,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:30522/aws-sdk-go-test-m3t66mrhlzlb16: dial tcp 192.168.122.45:30522: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#e47e4792-00f9-4fad-975f-b87243bf1cc5) |
<a name="e47e4792-00f9-4fad-975f-b87243bf1cc5"></a>
### 5568-c23b72e/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:31124
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp ceb64a5ef512:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-wg3pztat5xd1xo",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 362,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:31124/aws-sdk-go-test-wg3pztat5xd1xo: dial tcp 192.168.122.45:31124: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```

**Comment by harshavardhana** [comment]
Will review this tonight

File: cmd/server-main.go [hunk 1]
```diff
@@ -129,9 +129,8 @@ func serverHandleCmdArgs(ctx *cli.Context) {
 		fatalIf(checkPortAvailability(globalMinioPort), "Port %d already in use", globalMinioPort)
 	}
 
-	globalIsXL = (setupType == XLSetupType)
 	globalIsDistXL = (setupType == DistXLSetupType)
-	if globalIsDistXL {
+	if setupType == XLSetupType || globalIsDistXL {
```
Comment (harshavardhana) [inline cmd/server-main.go]: unnecessary change in this PR @donatello can you revert?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/server-main.go (hunk 1 above)
Comment (donatello) [reply cmd/server-main.go]: Done.

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |

**Comment by donatello** [comment]
Tested the PR with encrypted single-PUT objects as well - it works fine.

**Comment by harshavardhana** [comment]
> Tested the PR with encrypted single-PUT objects as well - it works fine.

will test this with PR #5576

**Review by harshavardhana — DISMISSED** [review]
LGTM tested with new SSE-C encryption PR and also with latest master with myriad of offsets and range GETs 

```go
        testCases := []struct {
                offset int64
                length int64
                md5    []byte
        }{
                {
                        offset: 10,
                        length: 10,
                },
                {
                        offset: 100,
                        length: 120,
                },
                {
                        offset: 64 * 1024,
                        length: 64 * 1024 * 1024,
                },
                {
                        offset: 100 * 1024 * 1024,
                        length: 100,
                },
                {
                        offset: 10 * 1024 * 1024,
                        length: 10 * 1024 * 1024,
                },
                {
                        offset: 2,
                        length: 64 * 1024 * 1024,
                },
                {
                        offset: 0,
                        length: 64*1024 + 1,
                },
                {
                        offset: 0,
                        length: 64*1024 - 1,
                },
                {
                        offset: 100 * 1024 * 1024,
                        length: 100 * 1024 * 1024,
                },
                {
                        offset: 100 * 1024 * 1024,
                        length: 1073741822,
                },
        }
```

File: cmd/erasure-readfile.go [hunk 2]
```diff
@@ -22,10 +22,79 @@ import (
...
+				}
+			} else {
+				successCount++
+				buffers[errVal.idx] = stageBuffers[errVal.idx]
+				stageBuffers[errVal.idx] = nil
+			}
+		}
+	}
+	if successCount == s.dataBlocks {
+		go drainErrors(launchIndex - finishedCount)
```
Comment (vadmeste) [inline cmd/erasure-readfile.go]: Isn't launchIndex always equal to finishedCount ?

File: cmd/erasure-readfile.go [hunk 3]
```diff
@@ -44,117 +113,120 @@ func (s ErasureStorage) ReadFile(writer io.Writer, volume, path string, offset,
...
-		writeLength := blocksize - startOffset
-		if length < writeLength {
-			writeLength = length
+		if needsReconstruction {
+			if err = s.ErasureDecodeDataBlocks(blocks); err != nil {
+				return f, errors.Trace(err)
+			}
+		}
+
+		var writeStart, writeLength int64
```
Comment (vadmeste) [inline cmd/erasure-readfile.go]: @donatello it is much easier to read if we do this style:
```
switch {
// First chunk
case chunkNumber == 0:
// Chunk in the middle
case chunkNumber > 0 && chunkNumber < numChunks-1:
...
// Last chunk
case chunkNumber == numChunks - 1:
....
}
```

What do you think ?

File: cmd/erasure-readfile.go (hunk 2 above)
Comment (donatello) [reply cmd/erasure-readfile.go]: Good point! I am further simplifying this.

File: cmd/erasure-readfile.go (hunk 3 above)
Comment (donatello) [reply cmd/erasure-readfile.go]: In case when `numChunks == 1` we would need to enter two cases in a switch statement like above - we would need to use fallthrough and it would not really be simpler. Let me just refactor this part slightly to improve its readability.

File: cmd/erasure-readfile.go [hunk 4]
```diff
@@ -44,117 +113,120 @@ func (s ErasureStorage) ReadFile(writer io.Writer, volume, path string, offset,
...
+				}
 			}
 		}
-		err = s.readConcurrent(volume, path, blockOffset, blocks, verifiers, errChans)
-		if err != nil {
-			return f, errors.Trace(errXLReadQuorum)
+
+		for i := range blocks {
+			if len(buffers[i]) != 0 {
+				_ = copy(blocks[i], buffers[i][chunkNumber*chunksize:])
```
Comment (vadmeste) [inline cmd/erasure-readfile.go]: Is there a way to avoid copying here, like we do `blocks[i] = buffers[i][chunkNumber*chunkSize:]` ? I guess to gain some performance.

Comment (harshavardhana) [reply cmd/erasure-readfile.go]: We also don't need to do `_ =` to avoid the return.

Comment (harshavardhana) [reply cmd/erasure-readfile.go]: Also we need to copy here @vadmeste

Comment (donatello) [reply cmd/erasure-readfile.go]: Ah ok, I will try.

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#08366f00-3321-4ab1-9243-02f4c70a062d) |
<a name="08366f00-3321-4ab1-9243-02f4c70a062d"></a>
### 5568-c2590c3/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:31697
ACCESS_KEY:      minioazure1
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp a8168c20abb6:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 1 seconds
(2/12) Running aws-sdk-php tests ... done in 4 minutes and 59 seconds
(3/12) Running aws-sdk-ruby tests ... done in 38 seconds
(4/12) Running awscli tests ... FAILED in 2 minutes and 55 seconds
{
  "name": "awscli",
  "duration": 8171,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it."
}

Executed 3 out of 12 tests successfully.
```

File: cmd/erasure-readfile.go (hunk 4 above)
Comment (donatello) [reply cmd/erasure-readfile.go]: @harshavardhana It looks like we can avoid copying.

Comment (donatello) [reply cmd/erasure-readfile.go]: @vadmeste @harshavardhana I've added this change to reduce copying in a separate commit to aid in reviewing - you can see the change more clearly I think. I can squash it too if thats preferable.

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |

**Review by harshavardhana — DISMISSED** [review]
LGTM

File: cmd/erasure-readfile.go [hunk 5]
```diff
@@ -44,117 +108,122 @@ func (s ErasureStorage) ReadFile(writer io.Writer, volume, path string, offset,
...
-		errChans[i] = make(chan error, 1)
+
+	chunksize := ceilFrac(blocksize, int64(s.dataBlocks))
+
+	// We read all whole-blocks of erasure coded data containing
+	// the requested data range.
+	//
+	// The start index of the erasure coded block containing the
+	// `offset` byte of data is:
+	partDataStartIndex := (offset / blocksize) * chunksize
```
Comment (vadmeste) [inline cmd/erasure-readfile.go:119]: @donatello can you move this block to a new function? There is really many declared variables here and it seems we need only to calculate partDataStartIndex & partDataLength, right ?

Comment (donatello) [reply cmd/erasure-readfile.go:119]: @vadmeste Ok, this is done - PTAL.

**Review by harshavardhana — APPROVED** [review]

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |

**Review by vadmeste — APPROVED** [review]
LGTM & tested
